### PR TITLE
Trust Spell Casting

### PR DESCRIPTION
**Pull request by zach2good**
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

**Since all the SQL changes were moved into another PR, I'll re-work this PR into being:**
- Trust spell list loading
- Trust spell selection
- Trust recast tracking
- A little ground work for Trust magic bursting

**As always, I have inflated the scope of this PR because I hate myself, and this now includes:**
- The bones of the ibm's gambit system, a basic behaviour definition system for trusts

### CODE REVIEW

**Comment by Omnione** [comment]
Pretty sure you know of the duplicate spells for Zeid causing travis to not complete, just leaving this here as a note.

**Comment by ibm2431** [comment]
Yell at me when you want to remove the WIP label~

(Any of us would work, Nyu and aether have the power too.)

File: src/map/spell.cpp [hunk 1]
```diff
@@ -163,6 +171,62 @@ bool CSpell::canHitShadow()
     return m_ID != SpellID::Meteor_II && canTargetEnemy();
 }
 
+SPELLFAMILY CSpell::getSpellFamily()
+{
+    // TODO: This should be in the db, hard coded for demo purposes!
```
Comment (TeoTwawki) [inline src/map/spell.cpp]: image

File: src/map/lua/lua_baseentity.cpp [hunk 2]
```diff
@@ -11744,7 +11744,7 @@ int32 CLuaBaseEntity::removeAmmo(lua_State* L)
 inline int32 CLuaBaseEntity::getWeaponSkillLevel(lua_State *L)
 {
     TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
-    TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC && m_PBaseEntity->objtype != TYPE_PET && m_PBaseEntity->objtype != TYPE_MOB);
+    TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC && m_PBaseEntity->objtype != TYPE_PET && m_PBaseEntity->objtype != TYPE_MOB && m_PBaseEntity->objtype != TYPE_TRUST);
```
Comment (TeoTwawki) [inline src/map/lua/lua_baseentity.cpp]: at this point maybe we should just debug break if its somehow an NPC lol

File: src/map/spell.cpp (hunk 1 above)
Comment (zach2good) [reply src/map/spell.cpp]: _it is done_

**Comment by zach2good** [comment]
**Note:**
Because I want/need the new db column, I've rebased off this PR:
#482

Until that goes in, this will be un-reviewable

File: scripts/globals/spells/trust/kupipi.lua [hunk 3]
```diff
@@ -22,19 +26,28 @@ function onSpellCast(caster, target, spell)
...
+    local FLASH  = 112
+    local ERASE  = 143
+    local DISPEL = 260
+
+    mob:addBehaviour(PARTY, HPP_LTE, 25, MA, SELECT_HIGHEST, tpz.magic.spellFamily.CURE)
+
+    mob:addBehaviour(PARTY, STATUS, tpz.effect.SLEEP_I, MA, SELECT_SPECIFIC, CURE_I)
+    mob:addBehaviour(PARTY, STATUS, tpz.effect.SLEEP_II, MA, SELECT_SPECIFIC, CURE_I)
+
+    mob:addBehaviour(PARTY, HPP_LTE, 75, MA, SELECT_HIGHEST, tpz.magic.spellFamily.CURE)
```
Comment (kaincenteno) [inline scripts/globals/spells/trust/kupipi.lua]: why is there two behaviors for CURE? shouldnt the one for HPP_LTE, 75 be enough? (sorry for my blunt question, just trying to provide feedback :cat: )

Comment (zach2good) [reply scripts/globals/spells/trust/kupipi.lua]: Useful question!

These behaviours are looked at in order behind the scenes, so if you wanted to weight a certain behaviour to be more important, you would put it higher on the list, or add another copy of it with different parameters. 

In this case, if someone in the party is critically low, Kupipi will try and drop a heal on them above all else. If not, she'll try to wake people up with Cure I, then she'll go to her regular type healing checks of 75% and all the other things in the list.

Comment (kaincenteno) [reply scripts/globals/spells/trust/kupipi.lua]: Awesome that makes sense @zach2good  :smile_cat: :  can't wait to try it out :smile_cat:

File: src/map/ai/helpers/gambits_container.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                    CStatusEffect* PSCEffect = target->StatusEffectContainer->GetStatusEffect(EFFECT_SKILLCHAIN, 0);
+                    std::list<SKILLCHAIN_ELEMENT> resonanceProperties;
+                    if (uint16 power = PSCEffect->GetPower())
+                    {
+                        resonanceProperties.push_back((SKILLCHAIN_ELEMENT)(power & 0xF));
+                        resonanceProperties.push_back((SKILLCHAIN_ELEMENT)(power >> 4 & 0xF));
+                        resonanceProperties.push_back((SKILLCHAIN_ELEMENT)(power >> 8));
+                    }
+
+                    // TODO: Move this to battleutils
```
Comment (ibm2431) [inline src/map/ai/helpers/gambits_container.cpp]: Or at least outside of the gambit checks which happen on every tick 😉 

Clarity edit: The resonance check obviously needs to happen on every gambit check, but at the moment the resonance _vector_ is being defined on each gambit tick

File: src/map/ai/helpers/gambits_container.cpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                    {
+                        // TODO: Get by element
+                        for (auto& spell : POwner->SpellContainer->m_damageList)
+                        {
+                            auto spell_element = spell::GetSpell(spell)->getElement();
+                            for (auto& chain_element : resonanceToElement[resonance_element])
+                            {
+                                if (spell_element == chain_element)
+                                {
+                                    spell_id = spell;
```
Comment (ibm2431) [inline src/map/ai/helpers/gambits_container.cpp:190]: Can pop a break in here

File: src/map/spell.cpp [hunk 6]
```diff
@@ -163,6 +163,14 @@ bool CSpell::isCure()
     return ((static_cast<uint16>(m_ID) >= 1 && static_cast<uint16>(m_ID) <= 11) || m_ID == SpellID::Cura || m_ID == SpellID::Cura_II || m_ID == SpellID::Cura_III);
 }
 
+bool CSpell::isDebuff()
+{
+    return ((getValidTarget() & TARGET_ENEMY) && getSkillType() == SKILL_ENFEEBLING_MAGIC) || m_ID == SpellID::Burn ||
+        m_ID == SpellID::Frost || m_ID == SpellID::Choke || m_ID == SpellID::Rasp || m_ID == SpellID::Shock ||
+        m_ID == SpellID::Drown || m_ID == SpellID::Stun || m_ID == SpellID::Curse || m_ID == SpellID::Bio ||
+        m_ID == SpellID::Bio_II || m_ID == SpellID::Bio_III || m_ID == SpellID::Bio_IV || m_ID == SpellID::Bio_V;
```
Comment (ibm2431) [inline src/map/spell.cpp]: This would be a good spot for those spell families~

File: src/map/ai/helpers/gambits_container.h [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    uint16 retry_delay;
+    time_point last_used;
+};
+
+class CGambitsContainer
+{
+public:
+    CGambitsContainer(CTrustEntity* trust)
+        : POwner(trust)
+    {}
```
Comment (ibm2431) [inline src/map/ai/helpers/gambits_container.h]: Are line breaks required here?

File: scripts/globals/spells/trust/kupipi.lua [hunk 8]
```diff
@@ -22,19 +26,28 @@ function onSpellCast(caster, target, spell)
...
+
+    mob:addGambit(PARTY, HPP_LTE, 75, MA, SELECT_HIGHEST, tpz.magic.spellFamily.CURE)
+
+    mob:addGambit(PARTY, NOT_STATUS, tpz.effect.PROTECT, MA, SELECT_HIGHEST, tpz.magic.spellFamily.PROTECTRA)
+    mob:addGambit(PARTY, NOT_STATUS, tpz.effect.SHELL, MA, SELECT_HIGHEST, tpz.magic.spellFamily.SHELLRA)
+
+    mob:addGambit(SELF, STATUS_FLAG, tpz.effectFlag.ERASABLE, MA, SELECT_SPECIFIC, ERASE)
+    mob:addGambit(PARTY, STATUS_FLAG, tpz.effectFlag.ERASABLE, MA, SELECT_SPECIFIC, ERASE)
+
+    mob:addGambit(TARGET, STATUS_FLAG, tpz.effectFlag.DISPELABLE, MA, SELECT_SPECIFIC, DISPEL)
```
Comment (ibm2431) [inline scripts/globals/spells/trust/kupipi.lua]: Does Kupipi dispel?

File: src/map/mob_spell_container.cpp [hunk 9]
```diff
@@ -85,6 +91,52 @@ void CMobSpellContainer::AddSpell(SpellID spellId)
...
+                matches.push_back(id);
+            }
+        };
+    };
+    searchInList(m_gaList);
+    searchInList(m_damageList);
+    searchInList(m_buffList);
+    searchInList(m_debuffList);
+    searchInList(m_healList);
+    searchInList(m_naList);
```
Comment (ibm2431) [inline src/map/mob_spell_container.cpp:133]: "Hey in the future we should only search lists that would have the appropriate spell."
"I know; that's why I want to give spell families parent families."

File: src/map/ai/controllers/trust_controller.cpp [hunk 10]
```diff
@@ -209,6 +216,23 @@ bool CTrustController::Ability(uint16 targid, uint16 abilityid)
     return false;
 }
 
+bool CTrustController::Cast(uint16 targid, SpellID spellid)
+{
+    FaceTarget(targid);
+    if (static_cast<CMobEntity*>(POwner)->PRecastContainer->Has(RECAST_MAGIC, static_cast<uint16>(spellid)))
+    {
+        return false;
+    }
```
Comment (ibm2431) [inline src/map/ai/controllers/trust_controller.cpp:225]: I know there's a check in getHighest, but are there checks for selectSpecific for if the trust has enough MP?

File: scripts/globals/gambits.lua [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+HPP_GTE          = 1
+MPP_LTE          = 2
+TP_GTE           = 3
+STATUS           = 4
+NOT_STATUS       = 5
+STATUS_FLAG      = 6
+NUKE             = 7
+SC_AVAILABLE     = 8
+NOT_SC_AVAILABLE = 9
+MB_AVAILABLE     = 10
```
Comment (ibm2431) [inline scripts/globals/gambits.lua]: Arrr... thar be global here!

**Review by ibm2431 — CHANGES_REQUESTED** [review]
Thanks as always, zach!

Note: I have sole free hold on this, so we can get this merged in relatively quickly~

File: scripts/globals/spells/trust/kupipi.lua (hunk 8 above)
Comment (zach2good) [reply scripts/globals/spells/trust/kupipi.lua]: Nope! I think I got confused when trying to put in as much functionality as possible. She does -na though, so I'll swap those in

**Review by ibm2431 — APPROVED** [review]

**Pull request merged**